websockets/src/websocket_connection.rs

```diff
@@ -63,7 +64,11 @@ impl WebSocketConn {
         self.sink.send(message).await.map_err(Into::into)
     }
 
-    pub(crate) async fn new(upgrade: Upgrade, config: Option<WebSocketConfig>) -> Self {
+    /// Create a `WebSocketConn` from an HTTP upgrade, with optional config and the specified role
+    ///
+    /// You should not typically need to call this; the trillium client and server both provide
+    /// your code with a `WebSocketConn`.
+    pub async fn new(upgrade: Upgrade, config: Option<WebSocketConfig>, role: Role) -> Self {
```

Comment:
While it'd be possible to make this API exempt from semver, doing so in a way that can't lead to breakage when upgrading one of the trillium crates and not others would require trillium-client to have a stricter dependency on trillium-websockets (e.g. an `=` dependency). That, in turn, would require a new version of trillium-client for each new version of trillium-websockets, even if trillium-client didn't otherwise need to change.  (With some care that could be "for each new minor version of trillium-websockets", post-1.0,  if handled carefully, but that's still a pain and and still seems likely to result in more new versions of trillium-client just to bump the version of trillium-websockets).

Given that, I'd prefer to handle this API via semver.

That said, we could mark it `doc(hidden)`, to make it more obvious that users don't need it. I'll do that.

Reply:
I'm convinced. The semver dependency between the two crates may eventually indicate introducing the more stable trillium-websockets-common crate you proposed, but there's no advantage to doing that until/unless necessary